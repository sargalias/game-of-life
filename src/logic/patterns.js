import { copyBoard } from "./board";

export const applyPattern = (board, pattern, xBoard, yBoard) => {
  // Returns a new board.
  // board: 2D array
  // pattern: 2D array
  const newBoard = copyBoard(board);
  for (let xPattern=0; xPattern<pattern.length; xPattern++) {
    for (let yPattern=0; yPattern<pattern[xPattern].length; yPattern++) {
      handleVal(newBoard, board, pattern, xBoard, yBoard, xPattern, yPattern);
    }
  }
  return newBoard
};

const handleVal = (newBoard, board, pattern, xBoard, yBoard, xPattern, yPattern) => {
  const patternVal = pattern[xPattern][yPattern];
  if (patternVal !== 0) {
    // index range check
    if (newBoard.length > xBoard + xPattern && newBoard[xBoard+xPattern].length > yBoard+yPattern) {
      const curVal = newBoard[xBoard+xPattern][yBoard+yPattern];
      const newVal = curVal === 0 ? pattern[xPattern][yPattern] : 0;
      newBoard[xBoard+xPattern][yBoard+yPattern] = newVal;
    }
  }
};
