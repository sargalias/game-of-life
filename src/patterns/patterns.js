export default {"34P13":[[0,0,0,0,1,1,1],[0],[0,0,0,0,0,0,1],[0,0,0,0,0,0,1,1],[0,0,0,0,0,1],[0,0,0,0,0,0,0,0,1],[0,0,0,0,1,0,0,0,1],[0,0,0,0,0,0,0,0,0,1,1],[1,1,0,0,1,1,1,0,0,0,0,0,1],[1,1,0,0,0,0,0,1,0,0,0,0,1,1,0,1],[0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,1],[0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,1],[0],[0],[0,0,0,0,0,0,1,1],[0,0,0,0,0,0,1,1]],"Single Cell":[[1]]}