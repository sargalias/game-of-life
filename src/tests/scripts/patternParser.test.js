const path = require('path');
const fs = require('fs');
const {
  patternParser,
  _parsePattern,
  _getAllFilenames,
  _readFile,
  _parseFile,
  _buildData,
  _writeData
} = require('../../../scripts/patterns/patternParser');
import { fileData, parsedData, filenames, allData, finalFile } from "./fixtures/patterns";

const inDir = path.join(__dirname, 'patterns_src');
const outPath = path.join(__dirname, 'patterns_out', 'patterns.js');
const outPath2 = path.join(__dirname, 'patterns_out', 'patterns2.js');


const removeWhitespace = (str) => {
  return str.replace(/\s/g, '');
};


test('_getAllFilenames works correctly', () => {
  expect.assertions(1);
  const expected = ['34p13_105.lif', 'simple_105.lif', 'single_cell_105.lif'];
  return _getAllFilenames(inDir)
    .then(files => {
      expect(files).toEqual(expected);
    });
});

test('_readFile works correctly', () => {
  const filepath = path.join(inDir, 'single_cell_105.lif');
  return _readFile(filepath).then((data) => {
    expect(removeWhitespace(data)).toBe(removeWhitespace(fileData[0]));
  });
});

test('_parsePattern works correctly with single cell pattern', () => {
  const data = fileData[0];
  expect(_parsePattern(data)).toEqual(parsedData[0]);
});

test('_parsePattern works correctly with simple cell pattern', () => {
  const data = fileData[1];
  expect(_parsePattern(data)).toEqual(parsedData[1]);
});

test('_parsePattern works correctly with normal cell pattern', () => {
  const data = fileData[2];
  expect(_parsePattern(data)).toEqual(parsedData[2]);
});

test('_parseFile should work correctly', () => {
  expect.assertions(1);
  const filename = path.join(inDir, 'simple_105.lif');
  return _parseFile(filename).then((data) => {
    expect(data).toEqual(parsedData[1]);
  });
});

test('_buildData should work correctly', () => {
  expect.assertions(1);
  return _buildData(inDir, filenames).then(allData => {
    expect(allData).toEqual({
      ...parsedData[0],
      ...parsedData[1],
      ...parsedData[2]
    });
  });
});

test('_writeData should work correctly', () => {
  expect.assertions(1);
  return _writeData(outPath, allData)
    .then(() => {
      return new Promise((resolve, reject) => {
        fs.readFile(outPath, 'utf8', (err, data) => {
          if (err) reject(err);
          resolve(data);
        });
      });
    }).then((data) => {
      expect(removeWhitespace(data)).toBe(removeWhitespace(finalFile));
    });
});

test('patternParser should work correctly', () => {
  return patternParser(inDir, outPath2).then(() => {
    return new Promise((resolve, reject) => {
      fs.readFile(outPath, 'utf8', (err, data) => {
        if (err) reject(err);
        resolve(data);
      });
    });
  }).then((data) => {
    expect(removeWhitespace(data)).toBe(removeWhitespace(finalFile));
  });
});
