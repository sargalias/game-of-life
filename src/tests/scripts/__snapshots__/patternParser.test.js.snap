// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildData should work correctly 1`] = `
Object {
  "34P13": Array [
    Array [
      0,
      0,
      0,
      0,
      1,
      1,
      1,
    ],
    Array [
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      1,
    ],
    Array [
      1,
      1,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
      0,
      0,
      0,
      1,
    ],
    Array [
      1,
      1,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      1,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      1,
    ],
    Array [
      0,
    ],
    Array [
      0,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      1,
    ],
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      1,
    ],
  ],
  "simple": Array [
    Array [
      1,
      0,
    ],
    Array [
      0,
      1,
    ],
  ],
  "single_cell": Array [
    Array [
      1,
    ],
  ],
}
`;

exports[`readFile works correctly 1`] = `
"#Life 1.05
#D Name: single_cell
#D The smallest known period 13 oscillator.
#D www.conwaylife.com/wiki/index.php?title=34P13
#N
#P -8 -8
*
"
`;
